### PR TITLE
PersistedStreamingMap SIGSEGV Fix

### PR DESCRIPTION
**Pull request by Maithem**
This patch fixes a bug where the iterator is accessed after it
has been closed, which ends up causing a segfault.

## Overview

Description:

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java [hunk 1]
```diff
@@ -40,6 +45,37 @@
...
+     *               Out of box, RocksDB will use LRU-based block cache
+     *               implementation with 8MB capacity.
+     * Index/Filter: Is a function of the block cache. Generally it infates
+     *               the block cache by about 50%. The exact number can be
+     *               retrieved via "rocksdb.estimate-table-readers-mem"
+     *               property.
+     * Write Buffer: Also known as memtable is defined by the ColumnFamilyOptions
+     *               option. The default is 64 MB.
+     */
+    public static Options getPersistedStreamingMapOptions() {
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Can you make all these parameters configurable? 
At least pass `Options` as a parameter in the constructor?
Or any other way that helps make all the parameters configurable

**Comment by vjeko** [comment]
Where was this being observed? AFAIK, this feature is not being used ATM.

**Comment by Maithem** [comment]
> Where was this being observed? AFAIK, this feature is not being used ATM.

Which feature are you referring to `PersistedStreamingMap` or `entryStream` ?

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java (hunk 1 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Ok.

**Comment by vjeko** [comment]
I am running your test on the master branch, and getting the following assertion error:

```
java.lang.IllegalStateException: Detected multi-threaded access to this iterator.

	at org.corfudb.runtime.collections.PersistedStreamingMap$RocksDbIterator.checkInvariants(PersistedStreamingMap.java:82)
	at org.corfudb.runtime.collections.PersistedStreamingMap$RocksDbIterator.hasNext(PersistedStreamingMap.java:104)
	at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1811)
	at java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
	at com.google.common.collect.Iterators$4.hasNext(Iterators.java:600)
	at org.corfudb.runtime.collections.CorfuTableTest.PersistedCorfuTableTests(CorfuTableTest.java:277)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.corfudb.AbstractCorfuTest$1$1.lambda$evaluate$0(AbstractCorfuTest.java:106)
	at java.lang.Thread.run(Thread.java:748)
```

Can you explain in more detail what the bug is?

**Comment by Maithem** [comment]
> I am running your test on the master branch, and getting the following assertion error:
> 
> ```
> java.lang.IllegalStateException: Detected multi-threaded access to this iterator.
> 
> 	at org.corfudb.runtime.collections.PersistedStreamingMap$RocksDbIterator.checkInvariants(PersistedStreamingMap.java:82)
> 	at org.corfudb.runtime.collections.PersistedStreamingMap$RocksDbIterator.hasNext(PersistedStreamingMap.java:104)
> 	at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1811)
> 	at java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
> 	at com.google.common.collect.Iterators$4.hasNext(Iterators.java:600)
> 	at org.corfudb.runtime.collections.CorfuTableTest.PersistedCorfuTableTests(CorfuTableTest.java:277)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
> 	at org.corfudb.AbstractCorfuTest$1$1.lambda$evaluate$0(AbstractCorfuTest.java:106)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> 
> Can you explain in more detail what the bug is?

There are two bugs, one that where `hasNext` closes the iterator and then the subsequent call fails with `IllegalStateException` (the bug which the Unit Test exercise). The second issue, is there is a race between closing the iterator and calling `status()`, It doesn't always produce an `AssertionError`, it can segfault. I'm guessing its bug in `RocksJava` that's related to `iterator.close()` and the GC

**Comment by vjeko** [comment]
> There are two bugs, one that where `hasNext` closes the iterator and then the subsequent call fails with `IllegalStateException` (the bug which the Unit Test exercise). The second issue, is there is a race between closing the iterator and calling `status()`, It doesn't always produce an `AssertionError`, it can segfault. I'm guessing its bug in `RocksJava` that's related to `iterator.close()` and the GC

How can there be a race? Race implies multi-threaded access. `PersistedStreamingMap` is meant to be consumed in a single threaded environment.

**Comment by vjeko** [comment]
> Which feature are you referring to `PersistedStreamingMap` or `entryStream` ?

I am referring to `PersistedStreamingMap`. The code path that this patch touches is not being used anywhere except for IDPS.

**Comment by Maithem** [comment]
> PersistedStreamingMap



> > There are two bugs, one that where `hasNext` closes the iterator and then the subsequent call fails with `IllegalStateException` (the bug which the Unit Test exercise). The second issue, is there is a race between closing the iterator and calling `status()`, It doesn't always produce an `AssertionError`, it can segfault. I'm guessing its bug in `RocksJava` that's related to `iterator.close()` and the GC
> 
> How can there be a race? Race implies multi-threaded access. `PersistedStreamingMap` is meant to be consumed in a single threaded environment.

There can be a race between a user thread and the GC thread. For example, the `RocksDbIterator` can have unsafe references to objects inside `RocksDB` and when calling methods on that iterator, it will try to deference and object that has been de-allocated (i.e. "finalized" by the GC).

**Comment by Maithem** [comment]
> > There are two bugs, one that where `hasNext` closes the iterator and then the subsequent call fails with `IllegalStateException` (the bug which the Unit Test exercise). The second issue, is there is a race between closing the iterator and calling `status()`, It doesn't always produce an `AssertionError`, it can segfault. I'm guessing its bug in `RocksJava` that's related to `iterator.close()` and the GC
> 
> How can there be a race? Race implies multi-threaded access. `PersistedStreamingMap` is meant to be consumed in a single threaded environment.

**Comment by Maithem** [comment]
> > Which feature are you referring to `PersistedStreamingMap` or `entryStream` ?
> 
> I am referring to `PersistedStreamingMap`. The code path that this patch touches is not being used anywhere except for IDPS.

This path can be invoked via https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java#L161

**Comment by vjeko** [comment]
> There can be a race between a user thread and the GC thread. For example, the `RocksDbIterator` can have unsafe references to objects inside `RocksDB` and when calling methods on that iterator, it will try to deference and object that has been de-allocated (i.e. "finalized" by the GC).

Maybe I am not fully understanding the use case. 

In order for the JVM to garbage collect `RocksDbIterator`, the object reference count has to be zero. However, if reference count is zero, this means that it is impossible for anyone to call `Iterator` methods since no one has a reference to it.

**Comment by Maithem** [comment]
> > There can be a race between a user thread and the GC thread. For example, the `RocksDbIterator` can have unsafe references to objects inside `RocksDB` and when calling methods on that iterator, it will try to deference and object that has been de-allocated (i.e. "finalized" by the GC).
> 
> Maybe I am not fully understanding the use case.
> 
> In order for the JVM to garbage collect `RocksDbIterator`, the object reference count has to be zero. However, if reference count is zero, this means that it is impossible for anyone to call `Iterator` methods since no one has a reference to it.

Not the  `RocksDbIterator` itself, but I was guessing some underlying resources. I haven't looked into the native code, but it seems like its encountering a segfault where it should throw an AssertionError. Possibly a race between underlying `dispose` (caused by close) and the `status`. Still need to investigate, but with this patch the issue doesn't seem to happen. 

Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  0x00007f58a40000f8
J 6323  org.rocksdb.RocksIterator.status0(J)V (0 bytes) @ 0x00007f59295d25e8 [0x00007f59295d2540+0xa8]
J 6322 C1 org.rocksdb.AbstractRocksIterator.status()V (30 bytes) @ 0x00007f592958475c [0x00007f5929584640+0x11c]
v  ~StubRoutines::call_stub
V  [libjvm.so+0x68825b]  JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0xddb
V  [libjvm.so+0x6d20d8]  jni_invoke_nonstatic(JNIEnv_*, JavaValue*, _jobject*, JNICallType, _jmethodID*, JNI_ArgumentPusher*, Thread*) [clone .constprop.110]+0x208
V  [libjvm.so+0x6d2fe5]  jni_CallVoidMethodA+0xd5
C  [libjdwp.so+0x1c014]  invoker_doInvoke+0xd94
C  [libjdwp.so+0x145ad]  reportEvents.part.2+0xad
C  [libjdwp.so+0x1498a]  event_callback+0x1da
C  [libjdwp.so+0x177f8]  cbBreakpoint+0xc8
V  [libjvm.so+0x772035]  JvmtiExport::post_raw_breakpoint(JavaThread*, Method*, unsigned char*)+0x315
V  [libjvm.so+0x67aedf]  InterpreterRuntime::_breakpoint(JavaThread*, Method*, unsigned char*)+0x1f
j  org.corfudb.runtime.collections.PersistedStreamingMap$RocksDbIterator.checkInvariants()V+36
J 6316 C1 org.corfudb.runtime.collections.PersistedStreamingMap$RocksDbIterator.hasNext()Z (122 bytes) @ 0x00007f5929b7772c [0x00007f5929b77660+0xcc]

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    /**
+     * Ensure that this iterator is still valid.
+     */
+    private void checkInvariants() {
+        try {
+            iterator.status();
+        } catch (RocksDBException e) {
+            throw new UnrecoverableCorfuError("There was an error reading the persisted map.", e);
```
Comment (hisundar) [inline runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: should the iterator be closed in case of an exception?

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: It doesn't matter because that exception should crash the process. On exit the resources should be cleaned.

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java [hunk 3]
```diff
@@ -318,9 +281,11 @@ public void clear() {
         // If it is nullptr, the iterator will read from an implicit snapshot as of the time the
         // iterator is created. The implicit snapshot is preserved by pinning resource.
         readOptions.setSnapshot(null);
-        final RocksIterator rocksIterator = rocksDb.newIterator(readOptions);
-        rocksIterator.seekToFirst();
-        return Streams.stream(new RocksDbIterator(rocksIterator));
+        final RocksDbEntryIterator entryIterator = new RocksDbEntryIterator(rocksDb.newIterator(readOptions), serializer);
+        Stream<Entry<K, V>> resStream = StreamSupport.stream(
+                Spliterators.spliteratorUnknownSize(entryIterator, Spliterator.ORDERED), false);
+        resStream.onClose(entryIterator::close);
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Stream consumers should use try-resource to invoke the close method.

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.AbstractMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+/**
+ *
+ * Created by Maithem on 1/21/20.
+ */
+
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: Please remove this and add a proper JavaDoc comment.

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        iterator.seekToFirst();
+        checkInvariants();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasNext() {
+        while (next == null && iterator.isValid()) {
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: Why are you looping here? Can you give a use-case?

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void close() {
+        // Release the underlying RocksDB resources
+        if (!iterator.isOwningHandle()) {
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: `close()` should be idempotent and an extra check can be done here to see if the iterator has been closed before. From the documentation:

```
     * However, implementers of this interface are strongly encouraged
     * to make their {@code close} methods idempotent.
```

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+            return res;
+        } else {
+            throw new NoSuchElementException();
+        }
+    }
+
+    /**
+     * Ensure that this iterator is still valid.
+     */
+    private void checkInvariants() {
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: All methods should check for multi-threaded access.  Also, this method should be executed on every API call.

File: test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java [hunk 8]
```diff
@@ -245,4 +255,47 @@ public void snapshotInvariant() {
         final Stream<Map.Entry<Integer, Integer>> result = map.entryStream();
         result.forEach(e -> map.put(new Random().nextInt(), 0));
     }
+
+    @Test
+    public void PersistedCorfuTableTests() {
```
Comment (vjeko) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: This test is using checkpointer specific logic, so it should be moved to `CheckpointTest`. A unit test should be checking for more specific functionality -- for example -- calling `hasNext()` on a closed iterator behaves as expected.

**Comment by vjeko** [comment]
> Not the `RocksDbIterator` itself, but I was guessing some underlying resources. I haven't looked into the native code

Can you share the core dump?

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java [hunk 9]
```diff
@@ -40,6 +45,37 @@
         RocksDB.loadLibrary();
     }
 
+    /**
+     * A set of options defined for disk-backed {@link CorfuTable}.
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Yes.

[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Changed.

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java (hunk 1 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Its kind of messy, because its passed to the constructor through a supplier. CorfuTable can be refactored in a different patch.

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java [hunk 10]
```diff
@@ -318,9 +281,11 @@ public void clear() {
         // If it is nullptr, the iterator will read from an implicit snapshot as of the time the
         // iterator is created. The implicit snapshot is preserved by pinning resource.
         readOptions.setSnapshot(null);
-        final RocksIterator rocksIterator = rocksDb.newIterator(readOptions);
-        rocksIterator.seekToFirst();
-        return Streams.stream(new RocksDbIterator(rocksIterator));
+        final RocksDbEntryIterator entryIterator = new RocksDbEntryIterator(rocksDb.newIterator(readOptions), serializer);
+        Stream<Entry<K, V>> resStream = StreamSupport.stream(
+                Spliterators.spliteratorUnknownSize(entryIterator, Spliterator.ORDERED), false);
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Not needed.

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java (hunk 4 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java (hunk 7 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java (hunk 6 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: It is
```
    public void close() {
        if (this.owningHandle_.compareAndSet(true, false)) {
            this.disposeInternal();
        }

    }
```

File: test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Maithem) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: Done.

Comment (Maithem) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: Its not checkpoint specific, although the checkpointer uses a CorfuTable that doesn't work.

File: test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java [hunk 11]
```diff
@@ -245,4 +255,47 @@ public void snapshotInvariant() {
...
+        for (List<Map.Entry<String, String>> partition : partitions) {
+            for (Map.Entry<String, String> entry : partition) {
+                result.put(entry.getKey(), entry.getValue());
+            }
+        }
+
+        assertThat(diskBackedMap.size()).isEqualTo(numKeys);
+        assertThat(result.size()).isEqualTo(numKeys);
+        for (Map.Entry<String, String> entry : result.entrySet()) {
+            assertThat(diskBackedMap.get(entry.getKey())).isEqualTo(diskBackedMap.get(entry.getKey()));
```
[the remark this answers is not in the document]
Comment (Maithem) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: Good catch.

File: test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java (hunk 8 above)
Comment (vjeko) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: This code was pretty much copied from `CheckpointWriter::appendObjectState`. Can you please add proper unit tests? Both positive and negative tests. You can add them to `DiskBackedCorfuClientTest`.

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java [hunk 12]
```diff
@@ -318,9 +274,13 @@ public void clear() {
...
         readOptions.setSnapshot(null);
-        final RocksIterator rocksIterator = rocksDb.newIterator(readOptions);
-        rocksIterator.seekToFirst();
-        return Streams.stream(new RocksDbIterator(rocksIterator));
+        final RocksDbEntryIterator entryIterator =
+                new RocksDbEntryIterator(rocksDb.newIterator(readOptions), serializer);
+        // It's necessary to use the RocksDbEntryIterator as try-with resource to ensure
+        // the iterator is close, otherwise underlying RocksDB resources can leak
+        Stream<Entry<K, V>> resStream = Streams.stream(entryIterator);
+        resStream.onClose(entryIterator::close);
```
Comment (WenbinZhu) [inline runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: According to this: https://github.com/facebook/rocksdb/blob/master/java/src/main/java/org/rocksdb/ReadOptions.java#L11
I think we need to call `readOptions.close()` before returning to prevent memory leak. (dispose is deprecated, so using close())

File: runtime/src/main/java/org/corfudb/runtime/collections/CheckedRocksIterator.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+public class CheckedRocksIterator implements RocksIteratorInterface {
+
+    private final RocksIterator iterator;
+    public CheckedRocksIterator(RocksIterator iterator) {
+        this.iterator = iterator;
+    }
+
+    private void checkHandle() {
+        if (!iterator.isOwningHandle()) {
+            throw new IllegalStateException("Handle is invalid");
```
[the remark this answers is not in the document]
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/CheckedRocksIterator.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java (hunk 12 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Didn't know about this. Thanks!

It will get released through finalized, but we shouldn't relay on that.

Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        this.serializer = serializer;
+        iterator.seekToFirst();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean hasNext() {
+        if (next == null && checkedIterator.isValid()) {
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: If we detect that the iterator has reached the end, we should go ahead and relinquish the resources immediately, rather than relying on `try-with-resource` or GC. This is a more robust solution since it is not guaranteed that the client will use `try-with-resource`.

File: test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java (hunk 8 above)
Comment (Maithem) [reply test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: This test belongs to CheckpointTest. Also, added more tests.

File: runtime/src/main/java/org/corfudb/runtime/collections/WrappedRocksIterator.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * A helper method that is called before accessing the iterator.
+     * This is necessary to make sure that access happens on an owned handle.
+     */
+    private void checkHandle() {
+        if (!iterator.isOwningHandle()) {
+            throw new IllegalStateException("RocksIterator handle is not valid");
+        }
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/WrappedRocksIterator.java]: This message should be more descriptive. The only was for the handle to be invalid is either if:

1) There is some other thread trying to access the iterator.
2) The iterator has been previously closed.

The message should reflect the above conditions.

File: runtime/src/main/java/org/corfudb/runtime/collections/WrappedRocksIterator.java [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        return res;
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void status() throws RocksDBException {
+        throw new UnsupportedOperationException("status");
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/WrappedRocksIterator.java]: `checkStatus()` and `status()` should behave the same. The implementation of this interface can be less restrictive, thus you are not required to throw `RocksDBException`, allowing you to merge these two methods.

File: runtime/src/main/java/org/corfudb/runtime/collections/WrappedRocksIterator.java [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * Verifies that the iterator is in a valid state.
+     */
+    private void checkStatus() {
+        try {
+            iterator.status();
```
Comment (vjeko) [inline runtime/src/main/java/org/corfudb/runtime/collections/WrappedRocksIterator.java]: You should document what constitutes a valid status. You can take a look at:

https://github.com/facebook/rocksdb/wiki/Iterator

And reference some parts of the Error Handling section.

File: test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java [hunk 18]
```diff
@@ -1,21 +1,31 @@
...
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 
 import org.assertj.core.api.Assertions;
 import org.assertj.core.data.MapEntry;
+import org.corfudb.runtime.CorfuRuntime;
 import org.corfudb.runtime.exceptions.unrecoverable.UnrecoverableCorfuError;
+import org.corfudb.runtime.object.ICorfuVersionPolicy;
 import org.corfudb.runtime.view.AbstractViewTest;
+import org.corfudb.util.serializer.Serializers;
```
Comment (vjeko) [inline test/src/test/java/org/corfudb/runtime/collections/CorfuTableTest.java]: None of these changes are necessary.

File: runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java (hunk 14 above)
Comment (Maithem) [reply runtime/src/main/java/org/corfudb/runtime/collections/RocksDbEntryIterator.java]: Not really, that assumes all stream consumers have to consume the full stream, which is an incorrect assumption. Consumers have to the stream whether explicitly or with try-with-resources
